codec.go

```diff
@@ -1,40 +1,100 @@
 package rpc
 
 import (
-	"github.com/ugorji/go/codec"
 	"io"
+
+	"github.com/ugorji/go/codec"
 )
 
 type decoder interface {
 	Decode(interface{}) error
 }
 
-type byteReadingDecoder interface {
-	decoder
-	io.ByteReader
+type encoder interface {
+	Encode(interface{}) <-chan error
```

Comment:
Might be out of scope for this PR, but one thing I've always found confusing is that encoder is *more* than just an encoder, and I always get lost trying to track down the *actual* Encode() call. What about renaming Encode to EncodeAndSend and Decode to ReadAndDecode or something?